### PR TITLE
171648183-respond-trip: respond to trip request

### PR DESCRIPTION
**Pull request by brianineza01**
#### Type of change

Please select the relevant option

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Others (cosmetics, styling, improvements)
- [ ] This change requires a documentation update

#### How Has This Been Tested?

- [x] Unit
- [ ] Integration
- [ ] End-to-end
- [ ] No testing required

#### How to Test
N/A
#### Any background context you want to add (Optional)
N/A
#### STORY ID

[Finishes #171648183]

### CODE REVIEW

**Review by Agnes-N — COMMENTED** [review]
you can add try and catch on this method of valideResponseTrip

> when there is an error it return it, so no need to implement try and catch

File: server/controllers/authController.js [hunk 1]
```diff
@@ -109,7 +109,6 @@ class AuthController {
           const {
             id, firstname, lastname, email, role, lineManager, username, isVerified
           } = user;
-
```
Comment (Irutingabo) [inline server/controllers/authController.js]: Please put back that empty line.

Comment (brianineza01) [reply server/controllers/authController.js]: feedback implemented

**Review by Irutingabo — APPROVED** [review]

File: server/controllers/tripController.js [hunk 2]
```diff
@@ -156,6 +156,39 @@ class TripController {
...
+   */
+  static async respondTrip(req, res) {
+    try {
+      const { tripId } = req.params;
+      const { status } = req.body;
+      const { role, id } = req.user;
+      const foundTrip = await TripHelper.findTripById(tripId, { role, id });
+      if (foundTrip) {
+        await TripHelper.updateTripById(tripId, { status });
+        const newTrip = await TripHelper.findTripById(tripId, { role, id });
```
Comment (placideirandora) [inline server/controllers/tripController.js]: Is this a new trip or an updated trip?

File: server/controllers/tripController.js [hunk 3]
```diff
@@ -156,6 +156,39 @@ class TripController {
...
+        return res.status(200).send({
+          status: 200,
+          data: newTrip
```
Comment (placideirandora) [inline server/controllers/tripController.js]: You should include a message telling the user about what has happened.

File: server/tests/trip.test.js [hunk 4]
```diff
@@ -200,4 +200,46 @@ describe('Trip test suite', () => {
         done(err);
       });
   });
+  it('13.should respond to a trip request', (done) => {
```
Comment (placideirandora) [inline server/tests/trip.test.js]: Kindly, remove the number and be explicit by using **should accept a trip request** instead.

File: server/tests/trip.test.js [hunk 5]
```diff
@@ -200,4 +200,46 @@ describe('Trip test suite', () => {
...
+      })
+      .end((err, res) => {
+        expect(res.body.status).to.equal(200);
+        expect(res.body).to.be.an('object');
+        expect(res.body).to.have.property('data');
+        done(err);
+      });
+  });
+  it('14.should not respond to a trip request as the status is wrong', (done) => {
+    router()
```
Comment (placideirandora) [inline server/tests/trip.test.js:218]: Here as well.

File: server/tests/trip.test.js [hunk 6]
```diff
@@ -200,4 +200,46 @@ describe('Trip test suite', () => {
...
+        status: 'no edit'
+      })
+      .end((err, res) => {
+        expect(res.body.status).to.equal(400);
+        expect(res.body).to.be.an('object');
+        expect(res.body).to.have.property('error');
+        done(err);
+      });
+  });
+  it('15.should not respond to a trip request as the user is not allowed to update the trip and the trip doesn\'t exist', (done) => {
```
Comment (placideirandora) [inline server/tests/trip.test.js]: Here as well.

File: swagger.json [hunk 7]
```diff
@@ -654,6 +654,60 @@
             "headers": {}
           }
         }
+      },
+      "patch": {
+        "description": "A line manager can respond to assigned trip requests",
+        "summary": "Respond to a trip request",
```
Comment (placideirandora) [inline swagger.json]: What do you mean by responding? That's implicit. Specify whether it's accepting or rejecting a trip request.

File: swagger.json [hunk 8]
```diff
@@ -654,6 +654,60 @@
...
+            "name": "tripId",
+            "type": "string",
+            "required": true,
+            "description": "tripId of a trip to be accepted or rejected"
+          },
+          {
+            "name": "Body",
+            "in": "body",
+            "required": true,
+            "description": "Super administrator is able to change roles of a particular user",
```
Comment (placideirandora) [inline swagger.json]: Have you worked on this as well?

**Review by placideirandora — CHANGES_REQUESTED** [review]

File: server/controllers/tripController.js (hunk 2 above)
Comment (brianineza01) [reply server/controllers/tripController.js]: I changed `newTrip` to `updatedTrip`

File: server/controllers/tripController.js (hunk 3 above)
Comment (brianineza01) [reply server/controllers/tripController.js]: I added it, the message

File: server/tests/trip.test.js (hunk 4 above)
Comment (brianineza01) [reply server/tests/trip.test.js]: feedback implemented

File: server/tests/trip.test.js (hunk 5 above)
Comment (brianineza01) [reply server/tests/trip.test.js:218]: feedback implemented

File: server/tests/trip.test.js (hunk 6 above)
Comment (brianineza01) [reply server/tests/trip.test.js]: feedback implemented

File: swagger.json (hunk 7 above)
Comment (brianineza01) [reply swagger.json]: feedback implemented

File: swagger.json (hunk 8 above)
Comment (brianineza01) [reply swagger.json]: feedback implemented

File: server/controllers/tripController.js [hunk 9]
```diff
@@ -156,6 +156,40 @@ class TripController {
...
+      const { tripId } = req.params;
+      const { status } = req.body;
+      const { role, id } = req.user;
+      const foundTrip = await TripHelper.findTripById(tripId, { role, id });
+      if (foundTrip) {
+        await TripHelper.updateTripById(tripId, { status });
+        const updatedTrip = await TripHelper.findTripById(tripId, { role, id });
+        return res.status(200).send({
+          status: 200,
+          message: `Trip is successfully ${status}`,
```
Comment (placideirandora) [inline server/controllers/tripController.js]: Kindly use **The Trip has been successfully updated** instead.

File: server/tests/trip.test.js (hunk 5 above)
Comment (placideirandora) [reply server/tests/trip.test.js:218]: It is obvious that the request will be rejected. Thus, update the description to **should reject a trip request when an invalid status is submitted**

File: server/tests/trip.test.js [hunk 10]
```diff
@@ -200,4 +200,46 @@ describe('Trip test suite', () => {
...
+        status: 'no edit'
+      })
+      .end((err, res) => {
+        expect(res.body.status).to.equal(400);
+        expect(res.body).to.be.an('object');
+        expect(res.body).to.have.property('error');
+        done(err);
+      });
+  });
+  it('should not accept or reject to a trip request as the user is not allowed to update the trip and the trip doesn\'t exist', (done) => {
```
Comment (placideirandora) [inline server/tests/trip.test.js]: You should be explicit. It is clear that the request will be rejected here. Thus, use **should reject the trip request when the user has no permission or the trip does not exist**.

File: swagger.json [hunk 11]
```diff
@@ -654,6 +654,60 @@
...
+            "name": "tripId",
+            "type": "string",
+            "required": true,
+            "description": "tripId of a trip to be accepted or rejected"
+          },
+          {
+            "name": "Body",
+            "in": "body",
+            "required": true,
+            "description": "line manager can accept or reject a trip request",
```
Comment (placideirandora) [inline swagger.json]: Is this the description of the body?

**Review by placideirandora — CHANGES_REQUESTED** [review]

File: server/tests/trip.test.js (hunk 5 above)
Comment (brianineza01) [reply server/tests/trip.test.js:218]: feedback implemented as you want

File: server/tests/trip.test.js (hunk 10 above)
Comment (brianineza01) [reply server/tests/trip.test.js]: feedback implemented

File: swagger.json (hunk 11 above)
Comment (brianineza01) [reply swagger.json]: Yes ofcourse,  It's the description of the body to be passed to the endpoint

File: server/tests/trip.test.js (hunk 10 above)
Comment (placideirandora) [reply server/tests/trip.test.js]: Not appropriately, remove the **Thus, use**. You have copy-pasted everything.

File: swagger.json [hunk 12]
```diff
@@ -654,6 +654,60 @@
...
+        "deprecated": false,
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [{
+            "name": "token",
+            "in": "header",
+            "required": true,
+            "type": "string",
+            "description": "token identify that a manager has signed in"
```
Comment (placideirandora) [inline swagger.json]: Use **token indicating** instead.

File: swagger.json (hunk 11 above)
Comment (placideirandora) [reply swagger.json:687]: Describe what should be passed in the body instead of repeating yourself. Because this has been already mentioned above.

**Review by placideirandora — CHANGES_REQUESTED** [review]

File: swagger.json (hunk 12 above)
Comment (brianineza01) [reply swagger.json]: I implemented it

File: server/controllers/tripController.js (hunk 9 above)
Comment (brianineza01) [reply server/controllers/tripController.js]: feedback implemented

File: server/tests/trip.test.js (hunk 10 above)
Comment (brianineza01) [reply server/tests/trip.test.js]: Sorry for that!!

File: swagger.json (hunk 11 above)
Comment (brianineza01) [reply swagger.json]: I already did that

**Review by placideirandora — APPROVED** [review]

**Pull request merged**